src/compiler/checker.ts

```diff
@@ -3874,28 +3910,25 @@ namespace ts {
             return links.resolvedType;
         }
 
-        function addTypeToSortedSet(sortedSet: Type[], type: Type) {
-            if (type.flags & TypeFlags.Union) {
-                addTypesToSortedSet(sortedSet, (<UnionType>type).types);
+        function addTypeToSet(typeSet: Type[], type: Type, typeKind: TypeFlags) {
+            if (type.flags & typeKind) {

```

Comment:
It's a bit suspicious to test bit flags with the `===` operator. If you ever include some other unrelated flag the test would fail.